src/arch/arm64/vm.c

```diff
@@ -0,0 +1,462 @@
+} arm64_kernel_header_t;
+
+int vm_arch_load_image(vm_t *v, void *data, size_t datasz)
+{
+    vm_arch_priv_t *priv = (vm_arch_priv_t *) v->priv;
+
+    arm64_kernel_header_t *header = data;
+    if (header->magic != 0x644d5241U)
+        return throw_err("Invalid kernel image\n");
+
+    uint64_t offset;
+    if (header->image_size == 0)
+        offset = 0x80000;
+    else
+        offset = header->text_offset;
+
+    if (offset + datasz >= ARM_KERNEL_SIZE ||
+        offset + header->image_size >= ARM_KERNEL_SIZE) {
+        return throw_err("Image size too large\n");
+    }
+
+    void *dest = vm_guest_to_host(v, ARM_KERNEL_BASE + offset);
+    memmove(dest, data, datasz);
+    priv->entry = ARM_KERNEL_BASE + offset;
+    return 0;
+}
+
+int vm_arch_load_initrd(vm_t *v, void *data, size_t datasz)
+{
+    vm_arch_priv_t *priv = (vm_arch_priv_t *) v->priv;
+    void *dest = vm_guest_to_host(v, ARM_INITRD_BASE);
+    memmove(dest, data, datasz);
+    priv->initrdsz = datasz;
+    return 0;
+}
+
+/* MPIDR is used by fdt generation.
+   Reference:
+ * https://developer.arm.com/documentation/ddi0601/2022-03/AArch64-Registers/MPIDR-EL1--Multiprocessor-Affinity-Register?lang=en
+ */
+#define ARM_MPIDR_BITMASK 0xFF00FFFFFFUL
+#define ARM_MPIDR_REG_ID ARM64_SYS_REG(3, 0, 0, 0, 5)
+
+static int get_mpidr(vm_t *v, uint64_t *mpidr)
+{
+    struct kvm_one_reg reg;
+    reg.addr = (uint64_t) mpidr;
+    reg.id = ARM_MPIDR_REG_ID;
+
+    if (ioctl(v->vcpu_fd, KVM_GET_ONE_REG, &reg) < 0)
+        return throw_err("Failed to get MPIDR register\n");
+
+    *mpidr &= ARM_MPIDR_BITMASK;
+    return 0;
+}
+
+/* The phandle of interrupt controller */
+#define FDT_PHANDLE_GIC 1
+
+/* Definitions of PCI spaces in device tree.
+ * Reference:
+ * - PCI Bus Binding to: IEEE Std 1275-1994
+ *   https://www.devicetree.org/open-firmware/bindings/pci/pci2_1.pdf
+ */
+#define FDT_PCI_IO_SPACE 0x01000000L
+#define FDT_PCI_MMIO_SPACE 0x02000000L
+
+/* Definitions of interrupt mapping
+ * Reference:
+ * https://www.kernel.org/doc/Documentation/devicetree/bindings/interrupt-controller/arm%2Cgic.txt
+ *
+ * Interrupt is described as
+ * <SPI or PPI> <IRQ Number> <Edge or Level Triggered > */
+#define ARM_FDT_IRQ_TYPE_SPI 0
+#define ARM_FDT_IRQ_TYPE_PPI 1
+#define ARM_FDT_IRQ_EDGE_TRIGGER 1
+#define ARM_FDT_IRQ_LEVEL_TRIGGER 4
+
+/* Helper macro to simplify error handling */
+#define __FDT(action, ...)                                                  \
+    do {                                                                    \
+        int __ret = fdt_##action(fdt, ##__VA_ARGS__);                       \
+        if (__ret >= 0)                                                     \
+            break;                                                          \
+        return throw_err("Failed to create device tree:\n %s\n %s\n",       \
+                         "_FDT(" #action __VA_OPT__(", ") #__VA_ARGS__ ")", \
```

Comment:
The message "_FDT(" #action __VA_OPT__(" is boring.
You can simply alert something like `fdt_property_cell` without mentioning `_FDT("

Reply:
Should I include the parameters passed to the fdt_... function in the error message, like `fdt_property_cell(fdt, .....)`? Or just simply mention the function name? `fdt_property_cell`

Reply:
I pushed a new revision. The error message contains both function name and the parameters.